src/Symfony/Component/Validator/Constraints/TargetAwareConstraintTrait.php

```diff
@@ -0,0 +1,37 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Validator\Constraints;
+
+use Symfony\Component\Validator\Constraint;
+
+/**
+ * This trait is intended as a helper for implementing TargetAwareConstraintInterface.
+ *
+ * Since the interface interface only makes sense for class constraints,
+ * the default targets is set to class constraint.
+ *
+ * @since 3.1
+ *
+ * @author Mathieu Lemoine <mlemoine@wemakecustom.com>
+ */
+trait TargetAwareConstraintTrait
+{
+    public $target;
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getTargets()
+    {
+        return Constraint::CLASS_CONSTRAINT;
```

Comment:
I see target awareness like a scope widening feature.

So called class constraints are applied to the class but usually have the validated instance as their scope (i.e. The constraint is checking that the current instance respects such and such invariant). A TA class constraint is using the class as its scope (i.e. The constraint is checking that the current instance does not break such and such invariant among all the other instances).

Following the same path, property constraints has the property as their scope (i.e. The constraint is checking that the current property respects such and such invariant). Widening the scope of such constraint to the current instance is already possible using a class constraint with a custom path.